### PR TITLE
Enable variable dimension (2d/3d)

### PR DESCRIPTION
**Pull request by streeve**
Because of the template on the input class, CabanaPD will now be a header-only library

Enable 2d or 3d simulations; primarily changes to the Particles class. Blocked by #57 which will simplify cell spacing here quite a bit. Also could block #33 since that is set up to be a 2d problem

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -2,17 +2,15 @@ configure_file(CabanaPD_config.hpp.cmakein CabanaPD_config.hpp @ONLY)
 
 file(GLOB HEADERS GLOB *.hpp)
 
-file(GLOB SOURCES *.cpp)
-
 install(FILES ${HEADERS} DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
 install(FILES ${CMAKE_CURRENT_BINARY_DIR}/CabanaPD_config.hpp DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
 
-add_library(CabanaPD ${SOURCES})
```
Comment (pabloseleson) [inline src/CMakeLists.txt:10]: @streeve : What does this change do?

File: src/CabanaPD_Particles.hpp [hunk 2]
```diff
@@ -396,6 +405,17 @@ class Particles<DeviceType, PMB, Dimension>
     friend class Comm<self_type, PMB>;
 
   protected:
+    template <class LocalMeshType, std::size_t Dim = dim>
+    std::enable_if_t<3 == Dim, void> dz( const LocalMeshType local_mesh,
+                                         const int zero[dim] )
+    {
+        dx[2] = local_mesh.measure( Cajita::Edge<Cajita::Dim::K>(), zero );
```
Comment (pabloseleson) [inline src/CabanaPD_Particles.hpp:412]: @streeve : Here it seems we are setting dz = 0. However, shouldn't this be done for the case of Dim == 2 instead?

File: src/CabanaPD_Solver.hpp [hunk 3]
```diff
@@ -327,13 +330,14 @@ class SolverElastic
     bool print;
 };
 
-template <class DeviceType, class ParticleType, class ForceModel,
-          class BoundaryCondition, class PrenotchType>
+template <class DeviceType, class InputType, class ParticleType,
```
Comment (pabloseleson) [inline src/CabanaPD_Solver.hpp:333]: @streeve : What is ParticleType?

File: src/CabanaPD_Solver.hpp [hunk 4]
```diff
@@ -327,13 +330,14 @@ class SolverElastic
...
+          class ForceModel, class BoundaryCondition, class PrenotchType>
 class SolverFracture
-    : public SolverElastic<DeviceType, ParticleType, ForceModel>
+    : public SolverElastic<DeviceType, InputType, ParticleType, ForceModel>
```
Comment (pabloseleson) [inline src/CabanaPD_Solver.hpp:336]: @streeve : What is InputType?

**Comment by pabloseleson** [comment]
@streeve : It seems here we are generalizing from 3D to a general dimension Dim. Correct? Note that when changing from 3D to 2D, one has to change the model, for example, from a fully 3D model to a 2D plane elasticity model with a change in the corresponding model constants.

File: src/CMakeLists.txt (hunk 1 above)
Comment (streeve) [reply src/CMakeLists.txt:10]: This makes it a header-only library (there are no source files anymore)

File: src/CabanaPD_Particles.hpp (hunk 2 above)
Comment (streeve) [reply src/CabanaPD_Particles.hpp:412]: This can be removed once the linked PR is merged

File: src/CabanaPD_Solver.hpp (hunk 3 above)
Comment (streeve) [reply src/CabanaPD_Solver.hpp:333]: `Particles`, but there are a number of different variants so it's easier to just template

File: src/CabanaPD_Solver.hpp (hunk 4 above)
Comment (streeve) [reply src/CabanaPD_Solver.hpp:336]: Same as above, `Inputs`, but now templated for the variable dimension

**Comment by streeve** [comment]
> Note that when changing from 3D to 2D, one has to change the model, for example, from a fully 3D model to a 2D plane elasticity model with a change in the corresponding model constants.

Agreed, feel free to add here if it's clear how to do so the force models. But since we don't have any explicit 2d examples right now we could merge that separately

File: src/CabanaPD_Particles.hpp (hunk 2 above)
Comment (streeve) [reply src/CabanaPD_Particles.hpp]: Fixed

**Review by pabloseleson — APPROVED** [review]

**Pull request merged**